File: src/main/kotlin/services/database/Schema.kt
@file:OptIn(ExperimentalTime::class)

package com.kitledger.services.database

import com.kitledger.domain.auth.SystemPermissionEnum
import kotlinx.serialization.json.Json
import org.jetbrains.exposed.v1.core.Column
import org.jetbrains.exposed.v1.core.Table
import org.jetbrains.exposed.v1.datetime.timestamp
import org.jetbrains.exposed.v1.json.jsonb
import java.util.*
import kotlin.time.ExperimentalTime

// Enums and Custom Types
enum class BalanceType {
    DEBIT, CREDIT
}

typealias BaseMetaProperty = Map<String, Any?>

// Tables

object AccountsTable : Table("accounts") {
    val id: Column<UUID> = uuid("id")
    val refId = varchar("ref_id", 64)
    val altId = varchar("alt_id", 64).nullable()
    val balanceType = customEnumeration(
        "balance_type",
        "varchar(10)",
        { value -> BalanceType.valueOf(value as String) },
        { BalanceType.valueOf(it.name) }
    )
    val ledgerId = uuid("ledger_id")
    val parentId = uuid("parent_id").nullable()
    val name = varchar("name", 64)
    val meta: Column<BaseMetaProperty?> = jsonb<BaseMetaProperty>("meta", Json).nullable()
    val active = bool("active").default(true)
    val createdAt = timestamp("created_at")
    val updatedAt = timestamp("updated_at")
    override val primaryKey = PrimaryKey(id)
}

object ApiTokensTable : Table("api_tokens") {
    val id: Column<UUID> = uuid("id")
    val userId = uuid("user_id")
    val name = varchar("name", 64)
    val createdAt = timestamp("created_at")
    val revokedAt = timestamp("revoked_at").nullable()
    override val primaryKey = PrimaryKey(id)
}

object EntityModelsTable : Table("entity_models") {
    val id: Column<UUID> = uuid("id")
    val refId = varchar("ref_id", 64)
    val altId = varchar("alt_id", 64).nullable()
    val name = varchar("name", 64)
    val active = bool("active").default(true)
    val createdAt = timestamp("created_at")
    val updatedAt = timestamp("updated_at")
    override val primaryKey = PrimaryKey(id)
}

object LedgersTable : Table("ledgers") {
    val id: Column<UUID> = uuid("id")
    val refId = varchar("ref_id", 64)
    val altId = varchar("alt_id", 64).nullable()
    val name = varchar("name", 64)
    val description = varchar("description", 255).nullable()
    val unitModelId = uuid("unit_model_id")
    val active = bool("active").default(true)
    val createdAt = timestamp("created_at")
    val updatedAt = timestamp("updated_at")
    override val primaryKey = PrimaryKey(id)
}

object PermissionAssignmentsTable : Table("permission_assignments") {
    val id: Column<UUID> = uuid("id")
    val permissionId = uuid("permission_id")
    val userId = uuid("user_id").nullable()
    val roleId = uuid("role_id").nullable()
    val createdAt = timestamp("created_at")
    val updatedAt = timestamp("updated_at")
    override val primaryKey = PrimaryKey(id)
}

object PermissionsTable : Table("permissions") {
    val id: Column<UUID> = uuid("id")
    val name = varchar("name", 64)
    val description = varchar("description", 255).nullable()
    val createdAt = timestamp("created_at")
    val updatedAt = timestamp("updated_at")
    override val primaryKey = PrimaryKey(id)
}

object RolesTable : Table("roles") {
    val id: Column<UUID> = uuid("id")
    val name = varchar("name", 64)
    val description = varchar("description", 255).nullable()
    val createdAt = timestamp("created_at")
    val updatedAt = timestamp("updated_at")
    override val primaryKey = PrimaryKey(id)
}

object SessionsTable : Table("sessions") {
    val id: Column<UUID> = uuid("id")
    val userId = uuid("user_id")
    val createdAt = timestamp("created_at")
    val expiresAt = timestamp("expires_at")
}

object SystemPermissionsTable : Table("system_permissions") {
    val id: Column<UUID> = uuid("id")
    val permission = enumerationByName<SystemPermissionEnum>("permission", 64)
    val userId = uuid("user_id")
    val createdAt = timestamp("created_at")
    val updatedAt = timestamp("updated_at")
    override val primaryKey = PrimaryKey(id)
}

object TransactionModelsTable : Table("transaction_models") {
    val id: Column<UUID> = uuid("id")
    val refId = varchar("ref_id", 64)
    val altId = varchar("alt_id", 64).nullable()
    val name = varchar("name", 64)
    val active = bool("active").default(true)
    val createdAt = timestamp("created_at")
    val updatedAt = timestamp("updated_at")
    override val primaryKey = PrimaryKey(id)
}

object UnitModelsTable : Table("unit_models") {
    val id: Column<UUID> = uuid("id")
    val refId = varchar("ref_id", 64)
    val altId = varchar("alt_id", 64).nullable()
    val name = varchar("name", 64)
    val active = bool("active").default(true)
    val baseUnitId = uuid("base_unit_id").nullable()
    val createdAt = timestamp("created_at")
    val updatedAt = timestamp("updated_at")
    override val primaryKey = PrimaryKey(id)
}

object UsersTable : Table("users") {
    val id: Column<UUID> = uuid("id")
    val firstName = varchar("first_name", 64)
    val lastName = varchar("last_name", 64)
    val email = varchar("email", 64)
    val passwordHash = text("password_hash")
    val createdAt = timestamp("created_at")
    val updatedAt = timestamp("updated_at")
    override val primaryKey = PrimaryKey(id)
}

object UserRolesTable : Table("user_roles") {
    val id: Column<UUID> = uuid("id")
    val userId = uuid("user_id")
    val roleId = uuid("role_id")
    val createdAt = timestamp("created_at")
    val updatedAt = timestamp("updated_at")
    override val primaryKey = PrimaryKey(id)
}

File: packages/core/src/schema.ts
import { relations } from "drizzle-orm";
import { boolean, index, jsonb, pgTable, text, timestamp, uuid, varchar } from "drizzle-orm/pg-core";

import { BalanceType } from "./accounts.js";

/**
 * Common database helper for timestamps
 */
export const timestamps = {
	created_at: timestamp("created_at", { mode: "date" }).defaultNow().notNull(),
	updated_at: timestamp("updated_at", { mode: "date" }),
};

export type BaseMetaProperty = Record<string, string | number | boolean | Date | null>;

/**
 * 1. Tables, Indexes, and Constraints
 * This allows us to defined and view the database schema in a structured way.
 */

/**
 * 1.1 Users and Roles
 */

export const api_tokens = pgTable(
	"api_tokens",
	{
		id: uuid("id").primaryKey(),
		user_id: uuid("user_id")
			.notNull()
			.references(() => users.id),
		name: varchar("name", { length: 64 }).notNull(),
		revoked_at: timestamp("revoked_at"),
	},
	(table) => [index("api_token_user_idx").on(table.user_id)],
);

export const permission_assignments = pgTable(
	"permission_assignments",
	{
		id: uuid("id").primaryKey(),
		permission_id: uuid("permission_id")
			.notNull()
			.references(() => permissions.id),
		user_id: uuid("user_id"),
		role_id: uuid("role_id"),
		...timestamps,
	},
	(table) => [
		index("permission_assignment_user_idx").on(table.user_id),
		index("permission_assignment_role_idx").on(table.role_id),
		index("permission_assignment_permission_idx").on(table.permission_id),
	],
);

export const permissions = pgTable("permissions", {
	id: uuid("id").primaryKey(),
	name: varchar("name", { length: 64 }).notNull().unique(),
	description: varchar("description", { length: 255 }),
	...timestamps,
});

export const roles = pgTable("roles", {
	id: uuid("id").primaryKey(),
	name: varchar("name", { length: 64 }).notNull().unique(),
	description: varchar("description", { length: 255 }),
	...timestamps,
});

export const sessions = pgTable("sessions", {
	id: uuid("id").primaryKey(),
	user_id: uuid("user_id")
		.notNull()
		.references(() => users.id),
	expires_at: timestamp("expires_at").notNull(),
	...timestamps,
});

export const system_permissions = pgTable(
	"system_permissions",
	{
		id: uuid("id").primaryKey(),
		permission: varchar("permission", { length: 64 }).notNull(),
		user_id: uuid("user_id")
			.notNull()
			.references(() => users.id),
		...timestamps,
	},
	(table) => [
		index("system_permission_user_idx").on(table.user_id),
		index("system_permission_permission_idx").on(table.permission),
	],
);

export const users = pgTable(
	"users",
	{
		id: uuid("id").primaryKey(),
		first_name: varchar("first_name", { length: 64 }).notNull(),
		last_name: varchar("last_name", { length: 64 }).notNull(),
		email: varchar("email", { length: 64 }).notNull().unique(),
		password_hash: text("password_hash").notNull(),
		...timestamps,
	},
	(table) => [index("user_email_idx").on(table.email)],
);

export const user_roles = pgTable("user_roles", {
	id: uuid("id").primaryKey(),
	user_id: uuid("user_id")
		.notNull()
		.references(() => users.id),
	role_id: uuid("role_id")
		.notNull()
		.references(() => roles.id),
	...timestamps,
});

/**
 * 1.5 Ledgers
 */
export const accounts = pgTable("accounts", {
	id: uuid("id").primaryKey(),
	ref_id: varchar("ref_id", { length: 64 }).notNull().unique(),
	alt_id: varchar("alt_id", { length: 64 }).unique(),
	balance_type: varchar("balance_type", { length: 10 }).notNull().$type<BalanceType>(),
	ledger_id: uuid("ledger_id").notNull(),
	parent_id: uuid("parent_id"),
	name: varchar("name", { length: 64 }).notNull(),
	meta: jsonb("meta").$type<BaseMetaProperty>(),
	active: boolean("active").default(true).notNull(),
	...timestamps,
});

export const ledgers = pgTable("ledgers", {
	id: uuid("id").primaryKey(),
	ref_id: varchar("ref_id", { length: 64 }).notNull().unique(),
	alt_id: varchar("alt_id", { length: 64 }).unique(),
	name: varchar("name", { length: 64 }).notNull(),
	description: varchar("description", { length: 255 }),
	active: boolean("active").default(true).notNull(),
	...timestamps,
});

/**
 * 2. Relations
 * These are for Drizzle query building features and have no direct effect on the database schema.
 */

export const apiTokenRelations = relations(api_tokens, ({ one }) => ({
	user: one(users, {
		fields: [api_tokens.user_id],
		references: [users.id],
	}),
}));

export const permissionRelations = relations(permissions, ({ many }) => ({
	permissionAssignments: many(permission_assignments),
	roles: many(roles),
	users: many(users),
}));

export const permissionAssignmentRelations = relations(permission_assignments, ({ one }) => ({
	permission: one(permissions, {
		fields: [permission_assignments.permission_id],
		references: [permissions.id],
	}),
	user: one(users, {
		fields: [permission_assignments.user_id],
		references: [users.id],
	}),
	role: one(roles, {
		fields: [permission_assignments.role_id],
		references: [roles.id],
	}),
}));

export const roleRelations = relations(roles, ({ many }) => ({
	users: many(user_roles),
	permissions: many(permission_assignments),
}));

export const systemPermissionRelations = relations(system_permissions, ({ one }) => ({
	user: one(users, {
		fields: [system_permissions.user_id],
		references: [users.id],
	}),
}));

export const userRelations = relations(users, ({ many }) => ({
	roles: many(user_roles),
	apiTokens: many(api_tokens),
	permissions: many(permission_assignments),
	system_permissions: many(system_permissions),
}));

export const userRoleRelations = relations(user_roles, ({ one }) => ({
	user: one(users, {
		fields: [user_roles.user_id],
		references: [users.id],
	}),
	role: one(roles, {
		fields: [user_roles.role_id],
		references: [roles.id],
	}),
}));
